### PR TITLE
Add getting specific user's score on a beatmap

### LINKED ISSUE #6330: Add api endpoint for retrieving player's score on a beatmap

### PR DESCRIPTION
**Pull request by notbakaneko**
new endpoint for user's score on beatmap
`beatmaps/:beatmap_id/scores/users/:user_id`; returns the `score` and its `position` in the ranking according to the mode and mods requested.

closes #6330

### CODE REVIEW

**Comment by nanaya** [comment]
There's a similar endpoint for multiplayer and it's using this format:

```
api/v2/rooms/{room}/playlist/{playlist}/scores/users/{user}
```

File: app/Http/Controllers/BeatmapsController.php [hunk 1]
```diff
@@ -64,20 +88,24 @@ public function scores($id)
...
+            }
 
-            if ($user !== null) {
-                $score = (clone $query)->where('user_id', $user->user_id)->first();
+            if ($userId !== null) {
+                $query->where('user_id', $userId);
             }
 
             $results = [
                 'scores' => json_collection($query->visibleUsers()->forListing(), 'Score', ['beatmap', 'user', 'user.country', 'user.cover']),
```
Comment (nanaya) [inline app/Http/Controllers/BeatmapsController.php:102]: `forListing` needs adjustment to not deduplicate user scores when requesting scores of specific user.

File: resources/docs/source/includes/_structures/beatmap_scores.md [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    "position": 1,
+    "score": {}
+  }
+}
+```
+
+Field              | Type              | Description
+------------------ | ----------------- | --------------------------------------------------------------------
+scores             | [Score](#score)[] | The list of top scores for the beatmap in descending order.
+userScore          | Object?           | The score of the current user. This is not returned if the current user does not have a score.
```
Comment (nanaya) [inline resources/docs/source/includes/_structures/beatmap_scores.md:15]: ...or when requesting scores of specific user.

File: resources/docs/source/includes/_structures/beatmap_scores.md [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+userScore.position | number            | The position of the user's score in the Beatmap's ranking, if available.
```
Comment (nanaya) [inline resources/docs/source/includes/_structures/beatmap_scores.md:16]: the whole `userScore` can probably be updated to just contain `json_item` of current user score with `rank_global` and `rank_country` includes... except we'll have to add support for passing mods for the transformer first. (also `rank_friend`. or maybe a whole new `rank` include which also returns which mod/mode it applies to)

File: resources/docs/source/includes/_structures/beatmap_scores.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+userScore.score    | [Score](#score)   | The details of the user's score in the Beatmap's ranking, if available
```
Comment (nanaya) [inline resources/docs/source/includes/_structures/beatmap_scores.md:17]: maybe we shouldn't document `userScore` but add `user_score` (or `current_user_score`) instead, and eventually remove this camelCased property.

**Review by nanaya — CHANGES_REQUESTED** [review]

File: app/Http/Controllers/BeatmapsController.php (hunk 1 above)
Comment (notbakaneko) [reply app/Http/Controllers/BeatmapsController.php:102]: for some reason I thought `forListing` did the sorting but it doesn't 🤔

File: routes/web.php [hunk 5]
```diff
@@ -361,6 +366,19 @@
...
+            Route::get('lookup', 'API\BeatmapsController@lookup');
+
+            Route::group(['prefix' => '{beatmap}'], function () {
+                Route::get('scores/users/{user}', 'BeatmapsController@userScore');
+                Route::get('scores', 'BeatmapsController@scores')->name('scores');
+            });
+        });
+
+
+
```
Comment (nanaya) [inline routes/web.php:379]: that's a lot of blank lines. Also around line 431 :point_down:

File: app/Http/Controllers/BeatmapsController.php [hunk 6]
```diff
@@ -86,4 +107,68 @@ public function scores($id)
...
+     *
+     * @queryParam mode The [GameMode](#gamemode) to get scores for.
+     * @queryParam mods An array of matching Mods, or none // TODO.
+     */
+    public function userScore($beatmapId, $userId)
+    {
+        $beatmap = Beatmap::findOrFail($beatmapId);
+        if ($beatmap->approved <= 0) {
+            abort(404);
+        }
```
Comment (nanaya) [inline app/Http/Controllers/BeatmapsController.php:135]: looks like can be combined as one query

File: app/Http/Controllers/BeatmapsController.php [hunk 7]
```diff
@@ -86,4 +107,68 @@ public function scores($id)
...
+
+        $params = get_params(request()->all(), null, [
+            'mode:string',
+            'mods:any',
```
Comment (nanaya) [inline app/Http/Controllers/BeatmapsController.php:139]: `string[]` here and `array_filter` down there? Apparently `presence` doesn't always return string :eyes:  (also for the one in `#scores`)

**Review by nanaya — CHANGES_REQUESTED** [review]

**Comment by Cristian-Sknz** [comment]
@nanaya Do you have any forecast of how long this Endpoint will be implemented?

**Review by nanaya — APPROVED** [review]

**Pull request merged**